docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,627 @@
+    * must contain only methods that satisfy TorchScript type-checking rules and are compilable to TorchScript IRs
+
+Example:
+
+Python classes can be used in TorchScript if they are annotated with @torch.jit.script, similar to how you would declare a TorchScript function:
+
+.. testcode::
+
+    @torch.jit.script
+    class MyClass:
+        def __init__(self, x: int):
+            self.x = x
+
+        def inc(self, val: int):
+            self.x += val
+
+
+Example:
+
+A TorchScript custom class type must "declare" all its instance attributes by assignments in __init__(). If  an instance attribute is not defined in __init__() but accessed in other methods of the class, the class cannot be compiled as a TorchScript class, as shown in the following example:
+
+.. testcode::
+
+    @torch.jit.script
+    class foo:
+        def __init__(self):
+            self.y = 1
+
+        # ERROR: self.x is not defined in __init__
+        def assign_x(self):
+            self.x = torch.rand(2, 3)
+
+ The above class will fail to be compiled and issue the following error:
+
+.. testoutput::
+
+    RuntimeError:
+    Tried to set nonexistent attribute: x. Did you forget to initialize it in __init__()?:
+    def assign_x(self):
+        self.x = torch.rand(2, 3)
+        ~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
+
+Example:
+
+In this example, a TorchScript custom class defines a class variable name, which is not allowed.
+
+.. testcode::
+
+    import torch
+
+    @torch.jit.script
+    class MyClass(object):
+        name = "MyClass"
+        def __init__(self, x: int):
+            self.x = x
+
+    def fn(a: MyClass):
+        return a.name
+
+It leads to the following compile-time error:
+
+.. testoutput::
+
+    RuntimeError:
+    Tried to access nonexistent attribute or method 'name' of type '__torch__.MyClass'. Did you forget to initialize an attribute in __init__()?:
+        File "test-class2.py", line 10
+    def fn(a: MyClass):
+        return a.name
+            ~~~~~~ <--- HERE
+
+Enum Type
+^^^^^^^^^^
+
+Like custom classes, semantics of enum type are user-defined and the entire class definition must be compilable to TorchScript IR and subject to TorchScript type-checking rules.
+
+::
+
+    TSEnumDef := "class" Identifier "(enum.Enum | TSEnumType)" ":"
+                ( MemberIdentifier "=" Value )+
+                ( MethodDefinition )*
+
+where
+
+    * Value must be TorchScript literals of type int, float, or str, and must be of the same TorchScript type
+    * TSEnumType is the name of a TorchScript enum type. Similar to Python enum, TorchScript allows restricted Enum subclassing, that is, subclassing an enum is allowed only if the enum does not define any members.
+    * MethodDefinition must be compilable to TorchScript IR and subject to TorchScript’s type checking rules
```

Comment:
Enums can not have any methods, they are ignored by TorchScript.

Reply:
Done